framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceImplementation.java

```diff
@@ -152,11 +155,51 @@ public void updateFromMethodInvocation(
       return;
     }
 
+    // Don't infer formal parameter types from recursive calls.
+    //
+    // When performing WPI on a library, if there are no external calls (only recursive calls), then
+    // each iteration of WPI would make the formal parameter types more restrictive, leading to an
+    // infinite (or very long) loop.
+    //
+    // Consider
+    //   void myMethod(int x) { ... myMethod(x-1) ... }`
+    // On one iteration, if x has type IntRange(to=100), the recursive call's argument has type
+    // IntRange(to=99).  If that is the only call to `MyMethod`, then the formal parameter type
+    // would be updated.  On the next iteration it would be refined again to @IntRange(to=98), and
+    // so forth.  A recursive call should never restrict a formal parameter type.
+    if (isRecursiveCall(methodInvNode)) {
+      return;
+    }
+
     List<Node> arguments = methodInvNode.getArguments();
     updateInferredExecutableParameterTypes(methodElt, arguments);
     updateContracts(Analysis.BeforeOrAfter.BEFORE, methodElt, store);
   }
 
+  /**
+   * Returns true if the given call is a recursive call.
+   *
+   * @param methodInvNode a method invocation
+   * @return true if the given call is a recursive call
+   */
+  private boolean isRecursiveCall(MethodInvocationNode methodInvNode) {
+    MethodTree enclosingMethod = TreePathUtil.enclosingMethod(methodInvNode.getTreePath());
+    if (enclosingMethod == null) {
+      return false;
+    }
+    ExecutableElement methodInvocEle = TreeUtils.elementFromUse(methodInvNode.getTree());
+    ExecutableElement methodDeclEle = TreeUtils.elementFromDeclaration(enclosingMethod);
+    if (!methodDeclEle.equals(methodInvocEle)) {
+      return false;
+    }
+
+    if (ElementUtils.isStatic(methodDeclEle)) {
```

Comment:
@smillst I don't think this is necessary (or desirable).  The above code has already established that the method is the same.  I think that inference should not be done for recursive calls, regardless of the value of the arguments (including the receiver).  I would change the above to `return methodDeclEle.equals(methodInvocEle);`.  Is there a reason this code is desirable?

Reply:
I agree  that this isn't necessary.  I was think of calls to the super method, but in that case the elements would not be equal. So, please make that change.